ksql-engine/src/main/java/io/confluent/ksql/analyzer/QueryAnalyzer.java

```diff
@@ -80,82 +90,86 @@ public AggregateAnalysis analyzeAggregate(final Query query, final Analysis anal
 
     aggregateAnalysis.setHavingExpression(
         ExpressionTreeRewriter.rewriteWith(aggregateExpressionRewriter,exp));
   }
 
-  private void processGroupByExpression(
+  private static void processGroupByExpression(
       final Analysis analysis,
       final AggregateAnalyzer aggregateAnalyzer
   ) {
     for (final Expression exp : analysis.getGroupByExpressions()) {
-      aggregateAnalyzer.process(exp, new AnalysisContext());
+      aggregateAnalyzer.processGroupBy(exp);
     }
   }
 
-  private void processSelectExpressions(
+  private static void processSelectExpressions(
       final Analysis analysis,
-      final AggregateAnalysis aggregateAnalysis,
+      final MutableAggregateAnalysis aggregateAnalysis,
       final AggregateAnalyzer aggregateAnalyzer,
       final AggregateExpressionRewriter aggregateExpressionRewriter
   ) {
     for (final Expression exp : analysis.getSelectExpressions()) {
-      processExpression(exp, aggregateAnalysis, aggregateAnalyzer);
+      aggregateAnalyzer.processSelect(exp);
 
       aggregateAnalysis.addFinalSelectExpression(
           ExpressionTreeRewriter.rewriteWith(aggregateExpressionRewriter, exp));
     }
   }
 
-  private static void processExpression(
-      final Expression expression,
-      final AggregateAnalysis aggregateAnalysis,
-      final AggregateAnalyzer aggregateAnalyzer) {
-    aggregateAnalyzer.process(expression, new AnalysisContext());
-    if (!aggregateAnalyzer.isHasAggregateFunction()) {
-      aggregateAnalysis.addNonAggResultColumns(expression);
+  private static void enforceAggregateRules(
+      final Query query,
+      final Analysis analysis,
+      final AggregateAnalysis aggregateAnalysis
+  ) {
+    if (!((QuerySpecification) query.getQueryBody()).getGroupBy().isPresent()) {
+      return;
     }
-    aggregateAnalyzer.setHasAggregateFunction(false);
-  }
 
-  private void enforceAggregateRules(final Query query, final AggregateAnalysis aggregateAnalysis) {
-    final Optional<GroupBy> groupBy = ((QuerySpecification) query.getQueryBody()).getGroupBy();
-    if (!groupBy.isPresent()) {
-      return;
+    if (aggregateAnalysis.getAggregateFunctions().isEmpty()) {
+      throw new KsqlException(
+          "GROUP BY requires columns using aggregate functions in SELECT clause.");
     }
 
-    final Set<Expression> groups = groupBy.get()
-        .getGroupingElements()
+    final Set<Expression> groupByExprs = ImmutableSet.copyOf(analysis.getGroupByExpressions());
+
+    final Set<Expression> unmatchedSelects = aggregateAnalysis.getNonAggregateSelectExpressions()
+        .entrySet()
         .stream()
-        .flatMap(group -> group.enumerateGroupingSets().stream())
-        .flatMap(Set::stream)
+        // Remove any that exactly match a group by expression:
+        .filter(e -> !groupByExprs.contains(e.getKey()))
+        // Remove any that are constants,
+        // or functions where all params exactly match a group by expression:
```

Comment:
nit: `or expressions where all column references are included as column references in the group by clause` would be more accurate right?

Reply:
Do we want to support functions where all params match a group by expression? (e.g. `CREATE TABLE OUTPUT AS SELECT LEN(SUBSTRING(source, 0, 2)), COUNT(*) FROM TEST GROUP BY SUBSTRING(source, 0, 2);`) I think that would be more complicated and out-of-scope of this PR.

Reply:
Yes to both.